### DISCUSSION TITLE
Creating fields / components using JSON/Dicts

### DISCUSSION THREAD

**Discussion by ToneseekerMusical** [TomSchimansky/CustomTkinter#1236] [Ideas]
I've been trying to work out how to make a dynamic UI that pulls layout information from a MongoDB database, and while it's possible to be done, I can't help but think that the method of coding CustomTkinter could be improved/extended.

Considering that MongoDB stores each document as a JSON file, it would be amazing to be able to create a layout using that raw data instead of having to hard code everything. Here's an example of how I see a tab component working:

```
self.json = {
  'Frame/Tab Name': {
    'placement': { 'padx': 5, 'pady': 5, 'sticky': 'ew' },
    'labels': {
      'placement': { 'padx': (5,40), 'pady': 5, },
      'label1': { 'type': ctk.CTkLabel, 'config': { 'text': 'Label Text' } },
      'label2': { 'type': ctk.CTkLabel, 'config': { 'text': 'Label Text' } },
      'label3': { 'type': ctk.CTkLabel, 'config': { 'text': 'Label Text' } },
      'label4': { 'type': ctk.CTkLabel, 'config': { 'text': 'Label Text' } },
    },
    'inputs': {
      'input1': { 'type': ctk.CTkOptionMenu, 'config': { 'values': values } },
      'input2': { 'type': ctk.CTkOptionMenu, 'config': { 'values': values } },
      'input3': { 'type': ctk.CTkOptionMenu, 'config': { 'values': values } },
      'input4': { 'type': ctk.CTkOptionMenu, 'config': { 'values': values } },
      'input5': { 'type': ctk.CTkEntry, 'config': { 'placeholder_text': 'Placeholder Text' } },
      'input6': { 'type': ctk.CTkEntry, 'config': { 'placeholder_text': 'Placeholder Text' } },
    }
  }
}

self.tab = ctk.CTkTabview( self.json )
self.frame = ctk.CTkFrame( self.json )
self.scrollFrame = ctk.CTkScrollframe( self.json )
```
The `labels` value in the dict should be optional, however, if it is present, and the number of inputs is greater than the number of labels, the inputs should wrap below the labels. This is assuming a two column, Label-Input layout. if there is no `labels` argument, a `columns` or a `rows` value should be provided in the `placement` dict that would determine how many rows/columns the tab or frame will display.

This would allow developers to easily implement access controls using the built in access control functions available in whatever database that we decide to use, and dynamically generate the application UI based on the user roles that have been set up in the database, and it would make it easy to pass the information collected back into the database. Ideally the labels and inputs would be put in their own dicts so that they can be queried individually when getting the data from the component as below:

```
for name, value in self.tab.inputs.items():
  if type( value ) == ctk.CTkLabel:
    print( f'{ name }: { value.cget("text") }' )
  if type( value ) in ( ctk.CTkEntry, ctk.CTkOptionMenu )
    print( f'{ name }: { value.get() }' )
```

**Comment by ElectricCandlelight** [comment]
Might not be my place to say but seems to be way outside the scope of this project. 

CustomTkinter / Tkinter is a toolkit / framework. What you're describing is more of an application.

You don't need to hardcode everything already so this could be implemented by yourself.

**Reply by ToneseekerMusical** [reply]
I'm suggesting adding a method to pass JSON/dicts to the Tabview, Frame, and Scrollframe objects that would be parsed into fields, not an entire application

**Comment by ElectricCandlelight** [comment]
Ah right I get you now. Still think it is outside of the project scope. 

Sounds like a good project to work on. 
There is something similar on GitHub [pytkgen](https://github.com/tmetsch/pytkgen) could be converted to CustomTkinter. There is probably a few others as well.
